pkg/detectors/gemini/gemini.go

```diff
@@ -0,0 +1,137 @@
+package gemini
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/detectorspb"
+)
+
+type Scanner struct{}
+
+const (
+	baseURL       = "https://api.gemini.com"
+	accountDetail = "/v1/account"
+	account       = "primary"
+)
+
+// Ensure the Scanner satisfies the interface at compile time.
+var _ detectors.Detector = (*Scanner)(nil)
+
+var (
+	client = common.SaneHttpClient()
+
+	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
+	keyPat    = regexp.MustCompile(`\b((?:master-|account-)[0-9A-Za-z]{20})\b`)
+	secretPat = regexp.MustCompile(`[A-Za-z0-9]{27,28}`)
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"master-", "account-"}
+}
+
+// FromData will find and optionally verify Gemini secrets in a given set of bytes.
+func (s Scanner) FromData(_ context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	idMatches := keyPat.FindAllStringSubmatch(dataStr, -1)
+	secretMatches := secretPat.FindAllStringSubmatch(dataStr, -1)
+
+	for _, match := range idMatches {
+		if len(match) != 2 {
+			continue
+		}
+		resMatch := strings.TrimSpace(match[1])
+
+		for _, secretMatch := range secretMatches {
+			resSecretMatch := strings.TrimSpace(secretMatch[0])
+
+			s1 := detectors.Result{
+				DetectorType: detectorspb.DetectorType_Gemini,
+				Raw:          []byte(resMatch),
+				RawV2:        []byte(resMatch + resSecretMatch),
+			}
+
+			if verify {
+				req, err := constructRequest(resSecretMatch, resMatch)
+				if err != nil {
+					continue
+				}
+
+				f := func() bool {
+					res, err := client.Do(req)
+					if err == nil {
+						defer res.Body.Close()
+						if res.StatusCode >= 200 && res.StatusCode < 300 {
+							s1.Verified = true
+							return true
+						}
+					}
+					return false
+				}
+				if !f() && detectors.IsKnownFalsePositive(resSecretMatch, detectors.DefaultFalsePositives, true) {
+					continue
+				}
```

Comment:
Okay, just to confirm my understanding.. assuming we find a false positive secret, we only want to add it to the list of results if it has been verified.

That being said, I think what we have now will add it to the list of results if `verify == false`, which I don't think we want? But I don't think we ever call `FromData` with `verify == false` anyway.

Reply:
Well all of this code is nested in an "if verify { ... }". So we would only be in here if it was true.